### PR TITLE
[#401] CLI: conf ls | get | set

### PR DESCRIPTION
**Pull request by ashu3103**
#### In Progress

- added support for new cli command - `conf ls`

#### TODO

- [x] add support for `conf get`
- [x] add support for `conf set`

@jesperpedersen PTAL

### CODE REVIEW

**Comment by jesperpedersen** [comment]
Yes, this looks correct.

Continue alongs these lines, add the `get` to `configuration.h` and transform the configuration into JSON (main and servers). Then do the filtering in `cli.c`

`set` will belong there too in the end

**Comment by jesperpedersen** [comment]
So, `fork()` in `main.c` with a call to `pgmoneta_configuration_get_json()` and so on

**Comment by jesperpedersen** [comment]
See `_STATUS` how to do it

**Comment by jesperpedersen** [comment]
@ashu3103 How is the `get` work going ?

**Comment by ashu3103** [comment]
> @ashu3103 How is the `get` work going ?

I am able to get the main and servers configuration to cli in json only the filtering part is remaining.

**Comment by jesperpedersen** [comment]
Great, sounds like you are making good progress !

**Comment by jesperpedersen** [comment]
No, `CONF_GET` only has the command - all filtering is done in `cli.c`

**Comment by ashu3103** [comment]
@jesperpedersen , I actually wanted a suggestion from you, After getting JSON configuration in `cli` we are stepping into `process_result` function (where read and print happens), so now the filtering should come somewhere in between reading and printing the response, in this case shall I create a new function `filter_process_result` (which read, filters and prints the response) or it has to be done in some other way?

**Comment by jesperpedersen** [comment]
@ashu3103 Yes, just create a new `filter_process_result` function that `CONF_GET` uses. `CONF_GET` is special anyway, because it will output the single result of the query, or an error.

So, read, transform and then filter is a good way forward

**Comment by jesperpedersen** [comment]
Yes, you do

```
  "Response": {
    "Configuration": {
      "pgmoneta": {
```

->

```
  "Response": {
    "Configuration": {
```

and

```
"primary": {
```

->

```
"Servers": {
      "primary": {
```

So, you would write

```
... conf get Servers.primary.host
```

to get that field.

Try and start with `pgmoneta_json_to_string(struct json* object, int32_t format, char* tag, int indent)` with `FORMAT_TEXT` and indent of `0`

@Jubilee101 Other ideas ?

**Comment by ashu3103** [comment]
@jesperpedersen PTAL

File: src/cli.c [hunk 1]
```diff
@@ -958,7 +987,7 @@ static void
 help_conf(void)
 {
    printf("Manage the configuration\n");
-   printf("  pgmoneta-cli conf [reload]\n");
+   printf("  pgmoneta-cli conf [reload | ls]\n");
```
Comment (jesperpedersen) [inline src/cli.c:990]: Add 'get'

File: src/libpgmoneta/conf.c [hunk 2]
```diff
@@ -0,0 +1,103 @@
+/*
```
Comment (jesperpedersen) [inline src/libpgmoneta/conf.c:1]: You can put this in `configuration.c`

File: src/libpgmoneta/configuration.c [hunk 3]
```diff
@@ -2012,6 +2021,146 @@ pgmoneta_reload_configuration(bool* restart)
...
+      pgmoneta_json_put(server_conf, CONFIGURATION_ARGUMENT_WORKERS, (uintptr_t)config->servers[i].workers, ValueInt64);
+      pgmoneta_json_put(server_conf, CONFIGURATION_ARGUMENT_BACKUP_MAX_RATE, (uintptr_t)config->servers[i].backup_max_rate, ValueInt64);
+      pgmoneta_json_put(server_conf, CONFIGURATION_ARGUMENT_NETWORK_MAX_RATE, (uintptr_t)config->servers[i].network_max_rate, ValueInt64);
+      pgmoneta_json_put(server_conf, CONFIGURATION_ARGUMENT_MANIFEST, (uintptr_t)config->servers[i].manifest, ValueInt64);
+      pgmoneta_json_put(server_conf, CONFIGURATION_ARGUMENT_TLS_CERT_FILE, (uintptr_t)config->servers[i].tls_cert_file, ValueString);
+      pgmoneta_json_put(server_conf, CONFIGURATION_ARGUMENT_TLS_CA_FILE, (uintptr_t)config->servers[i].tls_ca_file, ValueString);
+      pgmoneta_json_put(server_conf, CONFIGURATION_ARGUMENT_TLS_KEY_FILE, (uintptr_t)config->servers[i].tls_key_file, ValueString);
+      pgmoneta_json_put(server_conf, CONFIGURATION_ARGUMENT_EXTRA, (uintptr_t)config->servers[i].extra, ValueString);
+
+      pgmoneta_json_put(server_conf_j, config->servers[i].name, (uintptr_t)server_conf, ValueJSON);
```
Comment (jesperpedersen) [inline src/libpgmoneta/configuration.c:2145]: You need to use `json_append`

File: src/include/conf.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (jesperpedersen) [inline src/include/conf.h:1]: This can go into `configuration.h`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
You are missing some documentation - search for `git grep pgmoneta-cli`.

You are going in the right direction !

**Comment by fluca1978** [comment]
In case it helps, see this initial work for `conf get` <https://github.com/fluca1978/pgagroal/commit/ea07bd4eae1aa81c80da66ce05302501f57f4148>

**Comment by ashu3103** [comment]
> @ashu3103 Actually, thinking some more about it we can just move everything to top-level, like
> 
> ```
> "Response": {
>    "metrics": true,
>    ...
>    "primary": {
>       "host": "localhost",
>       ...
>    }
> }
> ```
> 
> Should be a simple change

I am a bit confused at this point, if we bring all the keys to the top level and if we query `conf get host`, the response will already have `host: localhost` (example) in its top-level. So, what exactly we expect in `Response`.

**Comment by ashu3103** [comment]
Also, can you tell me what is the need of bringing all the configuration to cli and do filtering at cli rather than at management handler of main.

**Comment by jesperpedersen** [comment]
This is the JSON reply from the server, so yes - you can do `pgmoneta-cli conf get host` and that will give you the value in `struct configuration`::`host` - e.g.

pgmoneta.conf
```
[pgmoneta]
host = mypgmoneta
...

[primary]
host = mypsql
...
```

```
pgmoneta-cli conf get host
mypgmoneta

pgmoneta-cli conf get primary.host
mypgsql
```

So, you just have to remove the `Configuration` and `Servers` nodes, and move the data "up"

**Comment by jesperpedersen** [comment]
Because then a client - other than pgmoneta-cli - can display the entire configuration with a single request

**Comment by ashu3103** [comment]
> This is the JSON reply from the server, so yes - you can do `pgmoneta-cli conf get host` and that will give you the value in `struct configuration`::`host` - e.g.
> 
> pgmoneta.conf
> 
> ```
> [pgmoneta]
> host = mypgmoneta
> ...
> 
> [primary]
> host = mypsql
> ...
> ```
> 
> ```
> pgmoneta-cli conf get host
> mypgmoneta
> 
> pgmoneta-cli conf get primary.host
> mypgsql
> ```
> 

Okay, but isn't that different from other commands which prints `Header`, `Request`, `Response` etc. (just confirming....)

> So, you just have to remove the `Configuration` and `Servers` nodes, and move the data "up"

Yeah, that's done :+1:

**Comment by jesperpedersen** [comment]
I think we will keep the output formats for `conf`, so

```
pgmoneta-cli -F text conf get encryption
none

pgmoneta-cli -F raw conf get encryption
0

pgmoneta-cli -F json conf get encryption
{"encryption": "none"}
```

**Comment by jesperpedersen** [comment]
The CLI will only use the `Response` part for the `conf` command - `Header` and `Request` won't be displayed. So, it will be like my example above

**Comment by ashu3103** [comment]
> I think we will keep the output formats for `conf`, so
> 
> ```
> pgmoneta-cli -F text conf get encryption
> none
> 
> pgmoneta-cli -F raw conf get encryption
> 0
> 
> pgmoneta-cli -F json conf get encryption
> {"encryption": "none"}
> ```

Yeah, this is better. Thanks!

File: src/cli.c [hunk 5]
```diff
@@ -1502,6 +1571,439 @@ process_result(SSL* ssl, int socket, int32_t output_format)
...
+   {
+      configuration_js = (struct json*)pgmoneta_json_get(response, context);
+   }
+
+   if (!configuration_js)
+   {
+      goto error;
+   }
+
+   if (!strcmp(key, "host"))
```
Comment (Jubilee101) [inline src/cli.c:1700]: This looks like a PITA.

An idea is that you can use iterator. Our json iterator works not only for json arrays but for json items as well. So just create an iterator and traverse the `configuration_js`, and put the string of the value into the response directly.

So something like:
```
if (!strcmp(key, iter->key) {
   pgmoneta_json_put(response, key, (uintptr_t)pgmoneta_value_to_string(iter->value), ValueString);
}
```
It'll save a lot of trouble for you. 

Sorry for proposing this 2 days later, after you've been through this... I've been busy with work recently.

Feel free to take the idea and see if you can make other parts simpler.

**Comment by jesperpedersen** [comment]
Note, #416

File: src/cli.c (hunk 5 above)
Comment (ashu3103) [reply src/cli.c]: @Jubilee101, This was really helpful. Thanks a lot!

**Comment by ashu3103** [comment]
@jesperpedersen PTAL

**Comment by jesperpedersen** [comment]
Currently we have

```
pgmoneta-cli -F text conf get encryption
encryption: none

pgmoneta-cli -F raw conf get encryption
{
  "encryption": 0
}

pgmoneta-cli -F json conf get encryption
{
  "encryption": 0
}

```

instead of

```
pgmoneta-cli -F text conf get encryption
none

pgmoneta-cli -F raw conf get encryption
0

pgmoneta-cli -F json conf get encryption
{"encryption": "none"}
```

For the first 2 we only need the value, not the key - and the last has to be a JSON_COMPACT format

File: src/include/configuration.h [hunk 6]
```diff
@@ -102,6 +164,31 @@ pgmoneta_validate_admins_configuration(void* shmem);
 int
 pgmoneta_reload_configuration(bool* restart);
 
+/**
+ * Get the json format of main configuration
+ * @param res The response of management
+ */
+void
+pgmoneta_add_configuration_response(struct json* res);
```
Comment (jesperpedersen) [inline src/include/configuration.h:172]: Can be `static int add_configuration_response()`

File: src/include/configuration.h [hunk 7]
```diff
@@ -102,6 +164,31 @@ pgmoneta_validate_admins_configuration(void* shmem);
...
+
+/**
+ * Get the json format of servers configuration
+ * @param res The response of management
+ */
+void
+pgmoneta_add_servers_configuration_response(struct json* res);
```
Comment (jesperpedersen) [inline src/include/configuration.h:179]: Can be `static int add_servers_configuration_response()`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
I think for `conf ls` will use the same infrastructure as `conf get`.

So, `conf ls` will output

```
pgmoneta-cli conf ls
pgmoneta_admins.conf
pgmoneta.conf
pgmoneta_users.conf
```

But, you can also use

```
pgmoneta-cli conf get admins_configuration_path
pgmoneta-cli conf get main_configuration_path
pgmoneta-cli conf get users_configuration_path
```

to get the same output. For `conf ls` `-F text` and `-F raw` will be the same, and `-F json` will use the JSON_COMPACT format

**Comment by ashu3103** [comment]
@jesperpedersen, PTAL

**Comment by jesperpedersen** [comment]
Do a run with Debug with clang

```
    #1 0x7f12738e28f1 in pgmoneta_append src/libpgmoneta/utils.c:1165:15
    #2 0x7f12738e2c5c in pgmoneta_append_char src/libpgmoneta/utils.c:1186:11
    #3 0x7f127383f54d in get_retention_string src/libpgmoneta/configuration.c:3238:19
    #4 0x7f127383e618 in add_servers_configuration_response src/libpgmoneta/configuration.c:2107:83
    #5 0x7f127383c687 in pgmoneta_conf_get src/libpgmoneta/configuration.c:2144:4
    #6 0x56073b150c04 in accept_mgt_cb src/main.c:1186:10
```

```
    #1 0x7f9cda65bb6d in pgmoneta_json_create src/libpgmoneta/json.c:373:21
    #2 0x55e9e3f6783b in get_config_key_result src/cli.c:1688:8
    #3 0x55e9e3f673cb in filter_process_result src/cli.c:1604:8
    #4 0x55e9e3f6557d in conf_get src/cli.c:1543:8
    #5 0x55e9e3f63c9f in main src/cli.c:841:22
```

**Comment by jesperpedersen** [comment]
`conf get` looks ok apart from the leaks, but `conf ls` is stitll the same

**Comment by ashu3103** [comment]
> `conf get` looks ok apart from the leaks, but `conf ls` is stitll the same

Ohh I forgot to make changes for ‘ls’, my bad
On it!

**Comment by jesperpedersen** [comment]
Once you are done with `conf get` then it is about `conf set` which is

```
pgmoneta-cli conf set key value
```

f.ex.

```
pgmoneta-cli conf set encryption aes
```

so that need to be translated before sent to the server to

```
Request: {
   "Key": "encryption",
  "Value": 0
}
``` 

and so on

Please, coordinate with @Jubilee101 in case there are something that can be improved...

**Comment by ashu3103** [comment]
> Do a run with Debug with clang
> 
> ```
>     #1 0x7f12738e28f1 in pgmoneta_append src/libpgmoneta/utils.c:1165:15
>     #2 0x7f12738e2c5c in pgmoneta_append_char src/libpgmoneta/utils.c:1186:11
>     #3 0x7f127383f54d in get_retention_string src/libpgmoneta/configuration.c:3238:19
>     #4 0x7f127383e618 in add_servers_configuration_response src/libpgmoneta/configuration.c:2107:83
>     #5 0x7f127383c687 in pgmoneta_conf_get src/libpgmoneta/configuration.c:2144:4
>     #6 0x56073b150c04 in accept_mgt_cb src/main.c:1186:10
> ```
> 
> ```
>     #1 0x7f9cda65bb6d in pgmoneta_json_create src/libpgmoneta/json.c:373:21
>     #2 0x55e9e3f6783b in get_config_key_result src/cli.c:1688:8
>     #3 0x55e9e3f673cb in filter_process_result src/cli.c:1604:8
>     #4 0x55e9e3f6557d in conf_get src/cli.c:1543:8
>     #5 0x55e9e3f63c9f in main src/cli.c:841:22
> ```

Fixed, PTAL

**Comment by jesperpedersen** [comment]
```
pgmoneta-cli -c pgmoneta.conf -F text conf ls
admin_configuration_path: ''
main_configuration_path: pgmoneta.conf
users_configuration_path: pgmoneta_users.conf

pgmoneta-cli -c pgmoneta.conf -F raw conf ls
{"admin_configuration_path":"","main_configuration_path":"pgmoneta.conf","users_configuration_path":"pgmoneta_users.conf"}

pgmoneta-cli -c pgmoneta.conf -F raw conf ls
{"admin_configuration_path":"","main_configuration_path":"pgmoneta.conf","users_configuration_path":"pgmoneta_users.conf"}

```

`-F text` and `-F raw` should only be the value -- they should both be the same

**Comment by ashu3103** [comment]
> ```
> pgmoneta-cli -c pgmoneta.conf -F text conf ls
> admin_configuration_path: ''
> main_configuration_path: pgmoneta.conf
> users_configuration_path: pgmoneta_users.conf
> 
> pgmoneta-cli -c pgmoneta.conf -F raw conf ls
> {"admin_configuration_path":"","main_configuration_path":"pgmoneta.conf","users_configuration_path":"pgmoneta_users.conf"}
> 
> pgmoneta-cli -c pgmoneta.conf -F raw conf ls
> {"admin_configuration_path":"","main_configuration_path":"pgmoneta.conf","users_configuration_path":"pgmoneta_users.conf"}
> ```
> 
> `-F text` and `-F raw` should only be the value -- they should both be the same, so
> 
> ```
> 
> pgmoneta.conf
> pgmoneta_users.conf
> ```

But that way how one's gonna figure out which file is for which configuration path?

**Comment by ashu3103** [comment]
> For `conf get` both `-F text` and `-F json` doesn't translate.
> 

Strange, the current logic is translating fields for me.

> `-F raw conf get` is ok

Can you send me the output at your end

**Comment by ashu3103** [comment]
Screenshot from 2024-11-18 23-10-17

**Comment by ashu3103** [comment]
> > Strange, the current logic is translating fields for me.
> 
> Check the `Header` - the `Compression` and `Encryption` fields

This must be at the `pgmoneta` side and not the `cli` side because we are not using translations at server's side only client side responses are translated

**Comment by jesperpedersen** [comment]
The `conf get XYZ` is ok with translation -- its the `conf get` command w/o the XYZ

**Comment by ashu3103** [comment]
> The `conf get XYZ` is ok with translation -- its the `conf get` command w/o the XYZ

Ohh, okay I'll fix that

**Comment by ashu3103** [comment]
@jesperpedersen, PTAL

**Comment by jesperpedersen** [comment]
It is weird,

```
-F text conf get
  Compression: none
  Encryption: none
```

```
-F raw
    "Compression": 2,
    "Encryption": 0,
```

```
-F json
    "Compression": "gzip",
    "Encryption": "none",
```

Where is `2` and `gzip` coming from ? I didn't specify a `-C` option...

**Comment by ashu3103** [comment]
@jesperpedersen Check now, actually I swapped compression and out_format while calling `pgmoneta_management_request_conf_get` and `pgmoneta_management_request_conf_ls`, my bad

**Comment by jesperpedersen** [comment]
@ashu3103 Yes, this looks good !

So, next is `set` which always `key` and `value` as arguments. Those get put into `ConfigKey` and `ConfigValue` in the `Request` part.

We could just output the input as the result - extracted from `Response` so

```
-F raw conf set entryption 0
0

-F text conf set entryption none
none

-F json conf set entryption none
{"encryption": "none"}
```

**Comment by ashu3103** [comment]
@jesperpedersen PTAL

File: src/main.c [hunk 8]
```diff
@@ -1149,6 +1149,66 @@ accept_mgt_cb(struct ev_loop* loop, struct ev_io* watcher, int revents)
...
+
+      pgmoneta_management_response_ok(NULL, client_fd, start_time, end_time, compression, encryption, payload);
+   }
+   else if (id == MANAGEMENT_CONF_GET)
+   {
+      pid = fork();
+      if (pid == -1)
+      {
+         pgmoneta_management_response_error(NULL, client_fd, server, MANAGEMENT_ERROR_STATUS_NOFORK, compression, encryption, payload);
+         pgmoneta_log_error("Status: No fork %s (%d)", server, MANAGEMENT_ERROR_STATUS_NOFORK);
```
Comment (jesperpedersen) [inline src/main.c:1174]: These need to be `_CONF_GET_` ones

File: src/main.c [hunk 9]
```diff
@@ -1149,6 +1149,66 @@ accept_mgt_cb(struct ev_loop* loop, struct ev_io* watcher, int revents)
...
+         pgmoneta_conf_get(NULL, client_fd, compression, encryption, pyl);
+      }
+   }
+   else if (id == MANAGEMENT_CONF_SET)
+   {
+      pid = fork();
+      if (pid == -1)
+      {
+         pgmoneta_management_response_error(NULL, client_fd, server, MANAGEMENT_ERROR_STATUS_NOFORK, compression, encryption, payload);
+         pgmoneta_log_error("Status: No fork %s (%d)", server, MANAGEMENT_ERROR_STATUS_NOFORK);
```
Comment (jesperpedersen) [inline src/main.c:1195]: These need to be `_CONF_SET_` ones

File: src/main.c [hunk 10]
```diff
@@ -1149,6 +1149,66 @@ accept_mgt_cb(struct ev_loop* loop, struct ev_io* watcher, int revents)
...
+         goto error;
+      }
+      else if (pid == 0)
+      {
+         struct json* pyl = NULL;
+         char* config_key = (char*)pgmoneta_json_get(request, MANAGEMENT_ARGUMENT_CONFIG_KEY);
+         char* config_value = (char*)pgmoneta_json_get(request, MANAGEMENT_ARGUMENT_CONFIG_VALUE);
```
Comment (jesperpedersen) [inline src/main.c:1202]: Extract those inside `pgmoneta_conf_set`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
```
pgmoneta-cli -c pgmoneta.conf -F json conf set unknown aes-256
```

need to return an error response

File: src/include/management.h [hunk 11]
```diff
@@ -158,6 +161,8 @@ extern "C" {
 #define MANAGEMENT_ARGUMENT_VALID                 "Valid"
 #define MANAGEMENT_ARGUMENT_WAL                   "WAL"
 #define MANAGEMENT_ARGUMENT_WORKERS               "Workers"
+#define MANAGEMENT_ARGUMENT_CONFIG_KEY            "ConfigKey"
+#define MANAGEMENT_ARGUMENT_CONFIG_VALUE          "ConfigValue"
```
Comment (jesperpedersen) [inline src/include/management.h:165]: Sorting

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgmoneta/configuration.c [hunk 12]
```diff
@@ -2032,6 +2038,1038 @@ pgmoneta_reload_configuration(bool* restart)
...
+
+   pgmoneta_json_destroy(payload);
+
+   pgmoneta_disconnect(client_fd);
+
+   pgmoneta_stop_logging();
+
+   exit(0);
+error:
+   pgmoneta_log_info("Inside main error");
```
Comment (jesperpedersen) [inline src/libpgmoneta/configuration.c:3062]: This need to be an `pgmoneta_log_error` and need to have context of why it failed

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by ashu3103** [comment]
@jesperpedersen PTAL

Also are there any docs which provide a descriptive idea about the error code to the client, if not can we add that too in this or another PR or we can add the description of the error code in the outcome field of the response itself.

Thoughts?

**Comment by jesperpedersen** [comment]
```
pgmoneta-cli conf get unknown
```

doesn't work

**Comment by jesperpedersen** [comment]
>Also are there any docs which provide a descriptive idea about the error code to the client, if not can we add that too in this or another PR or we can add the description of the error code in the Outcome field of the response itself.

We need a chapter in the manuals about it, but it is separate from this pull request I think

**Comment by ashu3103** [comment]
> ```
> pgmoneta-cli conf get unknown
> ```
> 
> doesn't work.
> 
> `-F text` and `-F raw` should give `Error`, `-F json` should give `{"Outcome": false}`

This can be only possible if we generate the error messages in `cli.c` because from main we will always get a response with whole json configuration and since we are filtering in `cli` we have to create error response explicitly.

**Comment by jesperpedersen** [comment]
I passed in `unknown` as the key, so that shouldn't resolve to anything

**Comment by ashu3103** [comment]
> I passed in `unknown` as the key, so that shouldn't resolve to anything

Yeah, so it will not be displaying anything except
`pgmoneta-cli: No connection to pgmoneta on /tmp/`

which is displayed because exit_code is 1.

**Comment by jesperpedersen** [comment]
Yes, but if you can't resolve `unknown` then the user still need a valid response

**Comment by ashu3103** [comment]
> Yes, but if you can't resolve `unknown` then the user still need a valid response

Understood, will be sticking with this format of output.
-F text and -F raw should give `Error`, -F json should give `{"Outcome": false}`

**Comment by jesperpedersen** [comment]
You can keep the exit code of 1 as well

**Comment by ashu3103** [comment]
> It says
> 
> ```
> Error
> pgmoneta-cli: No connection to pgmoneta on /tmp/
> ```
> 
> we don't need the last line -- only if there is no connection to pgmoneta

It will appear as long as we have `exit_code = 1` because of this section of code

```
   pgmoneta_disconnect(socket);

   if (configuration_path != NULL)
   {
      if (matched && exit_code != 0)
      {
         warnx("No connection to pgmoneta on %s", config->unix_socket_dir);
      }
   }
```

**Comment by jesperpedersen** [comment]
Better move that warnx to another part of the code

**Comment by jesperpedersen** [comment]
Log the warn when you know what it is... in this case we have a connection to pgmoneta - but the error was client side

**Comment by jesperpedersen** [comment]
If it isn't clear - just remove the code for now

**Comment by ashu3103** [comment]
> If it isn't clear - just remove the code for now

Okay so I'll comment it for now.

**Comment by ashu3103** [comment]
> If it isn't clear - just remove the code for now

Done

**Comment by jesperpedersen** [comment]
```
./pgmoneta-cli -c pgmoneta.conf -F XYZ conf set unknown aes
```

is still wrong

**Comment by ashu3103** [comment]
> ```
> ./pgmoneta-cli -c pgmoneta.conf -F XYZ conf set unknown aes
> ```
> 
> is still wrong

I am getting:
```
Header: 
  ClientVersion: 0.15.0
  Command: conf set
  Compression: none
  Encryption: none
  Output: text
  Timestamp: 20241124234310
Outcome: 
  Error: 2204
  Status: false
Request: 
  ConfigKey: unknown
  ConfigValue: aes
Response: 
  ServerVersion: 0.15.0
```

For conf set where the errors are coming from main, I am displaying the error in `Outcome`

**Comment by jesperpedersen** [comment]
```
./pgmoneta-cli -c pgmoneta.conf -F text conf set unknown aes
```

should give

```
Error
```

```
./pgmoneta-cli -c pgmoneta.conf -F raw conf set unknown aes
```

should give

```
Error
```

```
./pgmoneta-cli -c pgmoneta.conf -F json conf set unknown aes
```

should give

```
{"Outcome": false}
```

**Comment by ashu3103** [comment]
> ```
> ./pgmoneta-cli -c pgmoneta.conf -F text conf set unknown aes
> ```
> 
> should give
> 
> ```
> Error
> ```
> 
> ```
> ./pgmoneta-cli -c pgmoneta.conf -F raw conf set unknown aes
> ```
> 
> should give
> 
> ```
> Error
> ```
> 
> ```
> ./pgmoneta-cli -c pgmoneta.conf -F json conf set unknown aes
> ```
> 
> should give
> 
> ```
> {"Outcome": false}
> ```

okay

**Comment by jesperpedersen** [comment]
Rebased, and merged.

Thanks for your contribution !

**Pull request closed without merging**